File: 426. Convert Binary Search Tree to Sorted Doubly Linked List.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, left, right):
        self.val = val
        self.left = left
        self.right = right
"""
class Solution:
    def treeToDoublyList(self, root: 'Node') -> 'Node':
        if not root:
            return root
        dmhead = node = Node(0, None, None)
        ls = []
        self.inorder(root, ls)
        for item in ls:
            node.right = Node(item, node, None)
            node = node.right
        dmhead.right.left = node
        node.right = dmhead.right
        return dmhead.right
        
    def inorder(self, root, ls):
        if not root:
            return
        self.inorder(root.left, ls)
        ls.append(root.val)
        self.inorder(root.right, ls)

"""
# Definition for a Node.
class Node:
    def __init__(self, val, left, right):
        self.val = val
        self.left = left
        self.right = right
"""
class Solution:
    def treeToDoublyList(self, root: 'Node') -> 'Node':
        def inorder(root):
            nonlocal start, end
            if root:
                inorder(root.left)
                if end:
                    end.right = root
                    root.left = end
                else:
                    start = root
                end = root
                inorder(root.right)
        if not root:
            return
        start, end = None, None
        inorder(root)
        start.left = end
        end.right = start
        return start

File: 426. Convert Binary Search Tree to Sorted Doubly Linked List.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;

    public Node() {}

    public Node(int _val,Node _left,Node _right) {
        val = _val;
        left = _left;
        right = _right;
    }
};
*/
class Solution {
    Node first;
    Node last;
    
    public Node treeToDoublyList(Node root) {
        if (root==null) {
            return null;
        }
        helper(root);
        
        last.right = first;
        first.left = last;
        return first;
    }
    
    private void helper(Node node) {
        if (node != null) {
            helper(node.left);
            if (last != null) {
                last.right = node;
                node.left = last;
            } else {
                first = node;
            }
            last = node;
            helper(node.right);
        }
    }
}

class Solution {
    public Node treeToDoublyList(Node root) {
        Deque<Node> stack = new LinkedList<>();
        Deque<Node> dq = new LinkedList<>();
        if (root==null) {
            return root;
        }
        if (root.left == null && root.right == null) {
            root.left = root;
            root.right = root;
            return root;
        }
        Node curr = root, node = null, prev = null, head=null;
        while (curr != null || !stack.isEmpty()) {
            while(curr != null) {
                stack.push(curr);
                curr = curr.left;
            }
            curr = stack.pop();
            if (head==null) {
                head = curr;
            }
            prev = node;
            node = curr;
            node.left = prev;
            if (prev != null) {
                prev.right = node;
            }
            curr = curr.right;
        }
        node.right = head;
        head.left = node;
        return head;
    }
}